src/Npgsql/NpgsqlConnector.cs

```diff
@@ -1169,14 +1168,18 @@ internal ValueTask<IBackendMessage> ReadMessage(bool async, DataRowLoadingMode d
                             if (_originalTimeoutException != null)
                                 throw Break(_originalTimeoutException);
 
-                            // We have got a timeout while not reading the async notifications - trying to cancel a query
+                            // User requested the cancellation and it timed out
+                            if (_cancellationRequested)
```

Comment:
> There's a race condition here where the user calls NpgsqlCommand.Cancel just as the query reaches CommandTimeout, and we mistake the TimeoutException from the latter as the 2nd (hard) timeout for the former (because _cancellationRequested has already been set to true). The effect would be to skip CancellationTimeout and to break immediately, without giving PG cancellation a chance to work. I think the race condition was there before too (but am not sure anymore).

Yep, there is one 😢

> One way to fix this, is to save a timestamp for when cancellation will expire (instead of just the bool _cancellationRequested), and check if we've passed that.

It doesn't sound that much reliable.

> One unrelated thought is about what happens when cancel is called _after_ the 1st timeout has occurred (but before the 2nd). The current behavior is to ignore it (because cancellation has already been requested by the timeout), and that's probably fine - though we may want to always through OperationCanceledException if Cancel has been invoked, not TimeoutException. But this is really not important.

This one doesn't really matter. Cancellation just as well could happen after the second timeout has occured.